crypto11/crypto11.go

```diff
@@ -217,7 +253,7 @@ type PKCS11Config struct {
 // If config is nil, and the library has already been configured, the
 // context from the first configuration is returned (and
 // the error will be nil in this case).
-func Configure(config *PKCS11Config) (*pkcs11.Ctx, error) {
+func Configure(config *PKCS11Config, ctx PKCS11Context) (PKCS11Context, error) {
```

Comment:
@jmhodges - you mentioned that it's not ideal that we need to pass `nil` here in a few places. I don't think there's a good way to avoid adding this argument altogether (unless we want all the callers to need to create their own `pkcs11.Ctx` objects, which seems a bit onerous). I think the alternative you proposed was a wrapper that would only require the `PKCS11Config` that we would use from existing callers instead?

Reply:
Nil interface checks don't work exactly how you'd like them to work and so we avoid doing them: https://trstringer.com/go-nil-interface-and-interface-with-nil-concrete-value/ .

It seems like we're just trying to allow for a fake PKCS11Context to be injected. I would either:

Add a `ConfigureWithPKCS11Context(config, ctx)` function where the ctx is guaranteed by the caller to be non-nil. This is def the "always create their own Ctx objects" solution.

Or add a factory func type to the signature `Configure(config *PKCS11Config, factory func(*PKCS11Config) PKCS11Context)` with a known good Default factory func defined in the package that we can pass it that does the `pkcs11.New(config.Path)`.

It would be easier to see what was better if I knew what the GCPHSM code would look like. I suspect the factory one

Reply:
Ah, let me take back that "would be easier" line. I understand you're actually tyring to get the existing MakeKey under test. Sorry for the noise

Reply:
(I still think the factory one is probably preferable)

Reply:
Let me know what you think about the update I pushed - I think it's what you were aiming at?